harness/determined/searcher/search_runner.py

```diff
@@ -49,94 +34,210 @@ def run(
+                        first_event and last_event_id is not None and last_event_id < event.id
+                    )
+                    skip_event = (
+                        first_event and last_event_id is not None and last_event_id == event.id
+                    )
+                    first_event = False
+
+                    if skip_event:
+                        continue
+
+                    if event.initialOperations:
                         logging.info("initial operations")
                         operations = self.search_method.initial_operations()
-                    elif e.trialCreated:
-                        logging.info(f"trialCreated({e.trialCreated.requestId})")
-                        request_id = uuid.UUID(e.trialCreated.requestId)
+                    elif event.trialCreated:
+                        logging.info(f"trialCreated({event.trialCreated.requestId})")
+                        request_id = uuid.UUID(event.trialCreated.requestId)
                         self.search_method.searcher_state.trials_created.add(request_id)
                         self.search_method.searcher_state.trial_progress[request_id] = 0.0
                         operations = self.search_method.on_trial_created(request_id)
-                    elif e.trialClosed:
-                        logging.info(f"trialClosed({e.trialClosed.requestId})")
-                        request_id = uuid.UUID(e.trialClosed.requestId)
+                    elif event.trialClosed:
+                        logging.info(f"trialClosed({event.trialClosed.requestId})")
+                        request_id = uuid.UUID(event.trialClosed.requestId)
                         self.search_method.searcher_state.trials_closed.add(request_id)
                         operations = self.search_method.on_trial_closed(request_id)
-                    elif e.trialExitedEarly:
+                    elif event.trialExitedEarly:
                         # duplicate exit accounting already performed by master
                         logging.info(
-                            f"trialExitedEarly({e.trialExitedEarly.requestId},"
-                            f" {e.trialExitedEarly.exitedReason})"
+                            f"trialExitedEarly({event.trialExitedEarly.requestId},"
+                            f" {event.trialExitedEarly.exitedReason})"
                         )
-                        if e.trialExitedEarly.exitedReason is None:
+                        if event.trialExitedEarly.exitedReason is None:
                             raise RuntimeError(
                                 "trialExitedEarly event is invalid without exitedReason"
                             )
-                        request_id = uuid.UUID(e.trialExitedEarly.requestId)
-                        if e.trialExitedEarly.exitedReason in (
+                        request_id = uuid.UUID(event.trialExitedEarly.requestId)
+                        if event.trialExitedEarly.exitedReason in (
```

Comment:
there are potentially some sharp edges around handling for invalidHP.  i'll think more about this but i think otherwise i'm on board with this implementation for fault handling.

Reply:
also, it's probably the case that users would prefer to apply invalidHP filters in python as part of search method instead of relying on training code to raise invalidHP once we have custom searcher so this may be a moot point.